### PR TITLE
Bron-Kerbosch attestation aggregation

### PR DESCRIPTION
**Pull request by GeemoCandama**
## Issue Addressed

This branch will address #3733 

## Proposed Changes

Incorporate the Bron-Kerbosch algorithm that Satalia wrote into the attestation aggregation flow. More info in #3733

## Additional Info

@ankurdubey521 and I will be working on this.

### CODE REVIEW

**Comment by GeemoCandama** [comment]
I think it should be ready for a first pass.

Overview:
- Deleted the importing agg from naive_aggregation_pool since the unagg attestations should already be in the op_pool.
- During unagg attestation processing add them to the op_pool
- Made Compact Attestation Data Clone (curious if this is the best solution here"
- Altered fields of AttestationDataMap to a HashMap for aggregated_attestation and a HashMap for unaggregated_attestations.
- Altered methods on AttestationDataMap accordingly
- Removed greedy aggregation on insertion to the op_pool
- Port Bron-Kerbosch implementation from Satalia to Lighthouse
- Add `get_clique_aggregate_attestations_for_epoch -> Vec<(&CompactAttestationData, CompactIndexedAttestation<T>)>`  method (This is where most of the changes are. I'm curious if it would be better to return an Iterator)
- Use AttMaxCover of the output of the above as input to max_cover

Also the changes made a couple of previously used functions unused. Should I go ahead and delete them?

**Comment by GeemoCandama** [comment]
I think the general ideas are there however there are a few tests that stall forever that I'm working on:
```
test store_tests::delete_blocks_and_states has been running for over 60 seconds
test store_tests::prune_single_block_long_skip has been running for over 60 seconds
test store_tests::shuffling_compatible_simple_fork has been running for over 60 seconds
test store_tests::block_production_different_shuffling_early has been running for over 60 seconds
test store_tests::block_production_different_shuffling_long has been running for over 60 seconds
```
I also tried what I thought would be an optimization of going through the cliques and filtering them if they are a subset of another clique before aggregate the attestations. That didn't seem to improve the situation, so I reverted that for now. 

EDIT: This is no longer an issue

File: beacon_node/operation_pool/src/lib.rs [hunk 1]
```diff
@@ -215,28 +218,128 @@ impl<T: EthSpec> OperationPool<T> {
...
+                            {
+                                indices_to_remove.push(index);
+                                break;
+                            }
+                        }
+                    }
+                    for index in indices_to_remove.iter().rev() {
+                        clique_aggregates.remove(*index);
+                    }
+                    let mut clique_aggregates = clique_aggregates.into_iter();
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/lib.rs]: This line contains the bug!! It _moves_ `clique_aggregates` into an iterator, which means that later when we iterate by mutable reference (`for mut clique_aggregate in &mut clique_aggregates { .. }`) we actually _consume_ each attestation that we've iterated. That means whenever there are unaggregated attestations to add we throw away a large portion of our attestations! This also explains why nodes running with `--subscribe-all-subnets` are most affected!

I removed this line in favour of just iterating over the vec (which _won't_ consume the elements) and can confirm that it fixes the block profitability :tada:

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs]: This is the patch I applied:

```patch
diff --git a/beacon_node/operation_pool/src/lib.rs b/beacon_node/operation_pool/src/lib.rs
index 3daf1cd8b..e45e3e2e7 100644
--- a/beacon_node/operation_pool/src/lib.rs
+++ b/beacon_node/operation_pool/src/lib.rs
@@ -268,7 +268,6 @@ impl<T: EthSpec> OperationPool<T> {
                             res_att
                         })
                         .collect();
-
                     let mut indices_to_remove = Vec::new();
                     clique_aggregates
                         .sort_by(|a, b| a.attesting_indices.len().cmp(&b.attesting_indices.len()));
@@ -283,10 +282,10 @@ impl<T: EthSpec> OperationPool<T> {
                             }
                         }
                     }
+
                     for index in indices_to_remove.iter().rev() {
                         clique_aggregates.remove(*index);
                     }
-                    let mut clique_aggregates = clique_aggregates.into_iter();
 
                     // aggregate unaggregate attestations into the clique aggregates
                     // if compatible
@@ -299,7 +298,7 @@ impl<T: EthSpec> OperationPool<T> {
                             })
                         }) {
                             *num_valid += 1;
-                            for mut clique_aggregate in &mut clique_aggregates {
+                            for clique_aggregate in &mut clique_aggregates {
                                 if !clique_aggregate
                                     .attesting_indices
                                     .contains(&attestation.attesting_indices[0])
@@ -310,7 +309,8 @@ impl<T: EthSpec> OperationPool<T> {
                         }
                     }
 
-                    cliqued_atts.extend(clique_aggregates.map(|indexed| (data, indexed)));
+                    cliqued_atts
+                        .extend(clique_aggregates.into_iter().map(|indexed| (data, indexed)));
                 }
             }
             // include aggregated attestations from unaggregated attestations whose

```

**Comment by GeemoCandama** [comment]
attestation_packing
This is from the block packing dashboard. I collect the parallel iterator because the validty_fn is not Sync. I think some of this could be worked out. I tried to return an iterator from the function but the validity_fn closure is captured and cant be returned.

**Comment by GeemoCandama** [comment]
Before:
Screenshot from 2023-10-19 01-38-38
After:
Screenshot from 2023-10-19 01-37-54
It's a slight improvement on average. But After has slightly higher highs.

**Comment by michaelsproul** [comment]
Sorry for the small hijack. While reviewing the key change I noticed that there was a bug in the index math (the code was filtering before enumerating, so the indices ended up being off). I also removed double iteration which maybe could have been slowing things down.

Those changes are in this commit: 06aed57e08f41ab039125a99659aa54439119ae1.

I also spotted an opportunity to avoid some allocations in BK itself (9e781e1e6b57a05fe5242861bd8edbff93ac3b79). There might be more lurking in there that we could remove, but before we go tweaking that code too aggressively I think it would be good to have some more thorough tests. There are also a few unwraps & unsafe indexing in the BK code, which we need to guarantee are unreachable before merging (or we could remove them in favour of `Result`s). I was thinking we could use `quickcheck` to generate random graphs to run BK on, then check some invariants:

- The algorithm doesn't panic
- At least one clique is returned
- All the cliques are actual cliques (i.e. all vertices are connected to each other, for each returned clique).
- No clique is a subset of any other.

For the `is_compatible` function I think we could generate a random `HashSet` of adjacencies, and then the func can do `adjacencies.contains(&(i, j)) || adjacencies.contains(&(j, i))`.

I'd also be happy to work on this next week

File: beacon_node/operation_pool/src/attestation_storage.rs [hunk 2]
```diff
@@ -216,24 +217,58 @@ impl<T: EthSpec> AttestationDataMap<T> {
...
+                    indexed,
+                })
+            });
+
+        aggregates.chain(unaggregates)
     }
 
     pub fn stats(&self) -> AttestationStats {
         let mut stats = AttestationStats::default();
+        let mut data_to_num_attestations: HashMap<&CompactAttestationData, usize> = HashMap::new();
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/attestation_storage.rs]: I think we should keep this as simple as possible by excluding the unaggregated attestations from the `max_aggregates_per_data` calculation. Our packing algorithm still handles the aggregates separately from the unaggregates, so the former metric is still interesting.

If you wanted, we could add a metric for `max_unaggregates_per_data`, although I think we would also be OK without it, as its value would be quite predictable (approx `N / 32 / 64`).

File: beacon_node/operation_pool/src/lib.rs [hunk 3]
```diff
@@ -707,6 +854,15 @@ impl<T: EthSpec> OperationPool<T> {
     }
 }
 
+fn is_compatible<T: EthSpec>(
+    x: &&CompactIndexedAttestation<T>,
+    y: &&CompactIndexedAttestation<T>,
+) -> bool {
+    let x_attester_set: HashSet<u64> = x.attesting_indices.iter().cloned().collect();
+    let y_attester_set: HashSet<u64> = y.attesting_indices.iter().cloned().collect();
+    x_attester_set.is_disjoint(&y_attester_set)
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/lib.rs]: Rather than allocating hashsets here it will probably be faster to use the `signers_disjoint_from` method which uses bitfield math. It will still allocate a vec, but I've opened https://github.com/sigp/ssz_types/issues/14 to track the removal of that.

File: beacon_node/operation_pool/src/lib.rs [hunk 4]
```diff
@@ -1038,97 +1194,6 @@ mod release_tests {
         assert_eq!(op_pool.num_attestations(), committees.len());
     }
 
-    /// Adding lots of attestations that only intersect pairwise should lead to two aggregate
-    /// attestations.
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/lib.rs]: Can we try to bring this test back? I don't understand the fundamental reason why it would fail with the new aggregation, but maybe I'm missing something.

File: beacon_node/operation_pool/src/lib.rs [hunk 5]
```diff
@@ -215,28 +219,137 @@ impl<T: EthSpec> OperationPool<T> {
...
+                                indices_to_remove.push(index);
+                                break;
+                            }
+                        }
+                    }
+
+                    for index in indices_to_remove.iter().rev() {
+                        clique_aggregates.swap_remove(*index);
+                    }
+                    (data, clique_aggregates)
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/lib.rs]: I'm wondering two things about this:

1. Could we move the subset removals _before_ the signature aggregation step?
2. Would it be faster to loop `0..n x 0..n` without sorting, as this is $O(n^2)$ either way?

Maybe (1) is worth re-visiting later, as it might require us to restructure the output of `bron_kerbosch` to enable easier subset checking, e.g. by returning sets. I've been wondering if given the amount of popping and removing we do, whether we should use a persitent data structure for P, X and the output cliques in BK. Maybe something like:

https://docs.rs/rpds/latest/rpds/set/hash_trie_set/struct.HashTrieSet.html

**Review by michaelsproul — COMMENTED** [review]
This is looking good! Getting closer to being ready to merge I think!

File: beacon_node/operation_pool/src/lib.rs (hunk 4 above)
Comment (GeemoCandama) [reply beacon_node/operation_pool/src/lib.rs]: I think I may have misunderstood what that test was supposed to be
testing. 
I had thought this was about the storing of the attestations which
previously involved greedily aggregating. Based on your feedback I
assume I should go through the clique aggregation process to arrive
at two aggregate attestations. Is this correct?

File: beacon_node/operation_pool/src/lib.rs (hunk 5 above)
Comment (GeemoCandama) [reply beacon_node/operation_pool/src/lib.rs]: 1. We could definitely do that. Like I mentioned on Discord I think we could put off signature aggregation to after max cover even. But signature aggregation is a small part of the process where BK is the most compute consuming. Personally I am not sure what should be prioritized for this PR vs new PR's that improve this one.
2. When I tested with and without sorting, sorting was faster. I was able to reduce allocations without sorting but the algorithm was still faster. 

Interesting idea. Would this help avoid cloning also?

File: beacon_node/operation_pool/src/lib.rs (hunk 4 above)
Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs]: Yeah I think so. The optimal aggregation should be able to find the two aggregates formed by aggregating all the non-overlapping ones. The test might need some minor tweaks though

File: beacon_node/operation_pool/src/lib.rs (hunk 5 above)
Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs]: Yeah the persistent data structures should be cheaper to clone. But then they have higher overheads than Vec for indexing/iterating, so the total may be slower. Lets leave as-is for now and we can revisit later

File: beacon_node/operation_pool/src/lib.rs (hunk 4 above)
Comment (GeemoCandama) [reply beacon_node/operation_pool/src/lib.rs]: I added the test back but it's failing for now. I'm trying to figure out exactly why. The BK algo takes forever on the same size set of attestations. So I reduced it and it's producing 7739 cliques from a set of 32 attestations. After removing subsets this goes down to 17.

Comment (GeemoCandama) [reply beacon_node/operation_pool/src/lib.rs]: This makes more sense now and the way the attestations are currently constructed we should expect there to be more than 2 cliques.  BK produces multiple cliques that when paired with another fully covers the attester set. The speed of BK is concerning on even the size of the test. Is there an upper bound on the number of aggregated attestations for a particular AttestationData that we would have?

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs]: Usually there are around 16 aggregates per data, but it's random so could be up to 2x more than that quite often.

I agree the slow performance is concerning. It would be good to investigate this case and see if we can get it running a bit quicker. The persistent data structures (instead of lots of vec clones) might be the way to do it. Or it might just be that BK is slow sometimes due to its non-polynomial runtime.

Comment (GeemoCandama) [reply beacon_node/operation_pool/src/lib.rs]: I'm not sure if this test case was designed intentionally for this but the number of maximal cliques is much higher than for a graph with random edges. The minimum degree is N-2 since its a path complement graph. This makes things extra slow.

I have tried using the HashTrieSet but the way I've implemented it doesn't improve things on this test case in a noticeable way. 
Screenshot from 2023-10-29 11-05-49

File: beacon_node/operation_pool/src/bron_kerbosch.rs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+{
+    if p.is_empty() && x.is_empty() {
+        publish_clique(r);
+        return;
+    }
+
+    let pivot = find_pivot(&p, &x, neighbourhoods);
+    let pivot_neighbours: HashTrieSet<usize> = neighbourhoods[pivot].iter().cloned().collect();
+    
+    let ip = p.iter().cloned().filter(|e| !pivot_neighbours.contains(e)).collect::<HashTrieSet<_>>();
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/bron_kerbosch.rs]: The persistent data structures will work best if we can construct the updated ones as modifications of the existing ones. Doing `p.iter().cloned()` will make an entirely new data structure that doesn't share any memory with the previous one.

Instead, this could be something like:

```rust
let mut ip = p.clone();
for e in p.iter() {
    if pivot_neighbours.contains(e) {
       ip.remove_mut(e);
    }
}
```

This is probably a pattern we want to repeat, so we could make a standalone function (like `hash_set_filter`) which implements this.

File: beacon_node/operation_pool/src/lib.rs (hunk 4 above)
Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs]: I think it would be cool to see how fast we can get it, but agree we probably shouldn't block on it. I added some suggestions about the persistent data structures here: https://github.com/sigp/lighthouse/pull/4507#discussion_r1375713327

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs]: @GeemoCandama I spent some time playing with this test and my conclusion is that we need some way to avoid the exponential blow-up from clique enumeration. 

On graphs of ~70 aggregates in the `attestation_pairwise_overlapping` test I saw BK generate millions of cliques without terminating. The bound on the number of cliques is $O(3^{n/3})$, which for 70 aggregates is enormous (135B). For the realistic number of 16 aggregates it's fine, only ~350, but I would prefer that we don't embed an exponential-time algorithm without an escape hatch.


Some ideas I had were:

- Add a `limit` parameter to BK that aborts the algorithm once `limit` cliques have been produced.
- If there are more than `limit` aggregates for a single data just return the best one without calculating any cliques.

**Comment by michaelsproul** [comment]
I was testing the performance with `--subscribe-all-subnets --import-all-attestations` and noticed that it starts to deteriorate after some time. I think it's because we're iterating and processing ~900k unaggregated attestations every slot once the pool fills up.

attestation_packing

This is even with max-aggregates-per-data set to 1!

I realised that the `validity_filter` is somewhat unnecessary to run for every unaggregated attestation, as the filter we actually use only depends on the checkpoint & the attestation data. I made a hacky patch for this [here](https://github.com/michaelsproul/lighthouse/commit/b262f2745bcd48d8954c9e32ae0e0b48ff85a57d), which seems to have improved the performance. I'll clean that up tomorrow or early next week (or hand it over to you if you're keen). I think a better solution would be to make the validity filter take a pair of `Epoch` (or `CheckpointKey`) and `&AttestationData`.

I'll post an updated graph once it's had a chance to warm up for a bit. The time seems to be more like the 30ms we see for regular nodes :ok_hand:

**Comment by michaelsproul** [comment]
Looks a lot better :tada:

attestation_packing_fixed

**Comment by michaelsproul** [comment]
I think this is ready to go. I've been running it on one of our Holesky beacon nodes for the last few days without issue. If anything it's a little bit faster than `unstable`.

Here's the median attestation packing time for this PR (green) vs unstable (purple):

attestation_packing_time_median

Here's the 95th percentile:

attestation_packing_time_95pc

Here's total block production time (median over 12h):

total_time_median

95th percentile over 12h:

total_time_95pc

Investigating that spike with a shorter interval shows it's a spike to 2s+

total_time_1h_95pc

This doesn't really correspond to a spike in attestation packing time, so I suspect it's another issue (lord knows we have state issues while packing blocks). The version of `unstable` running on the other BNs is also running without the changes from https://github.com/sigp/lighthouse/pull/4794.
 
I'll let @paulhauner do a once-over of the Rayon stuff (because we've been bitten before), and then we can merge.

**Review by michaelsproul — APPROVED** [review]

**Comment by michaelsproul** [comment]
Logs from the spike also seem to show that it was actually only a spike to 1.2s, but even with histogram buckets I don't quite see how that makes any sense

> Nov 08 06:08:01.222 DEBG Processed HTTP API request              method: GET, path: /eth/v2/validator/blocks/293440, status: 200 OK, elapsed: 1.192266986s

File: beacon_node/operation_pool/src/lib.rs [hunk 7]
```diff
@@ -215,28 +235,121 @@ impl<T: EthSpec> OperationPool<T> {
...
+        let mut cliques_from_aggregates: Vec<_> = aggregate_attestations
+            .into_par_iter()
+            .filter(|(data, _)| {
+                data.slot + spec.min_attestation_inclusion_delay <= state.slot()
+                    && state.slot() <= data.slot + T::slots_per_epoch()
+            })
+            .map(|(data, aggregates)| {
+                let aggregates: Vec<&CompactIndexedAttestation<T>> = aggregates
+                    .iter()
+                    .filter(|_| validity_filter(checkpoint_key, data))
```
Comment (paulhauner) [inline beacon_node/operation_pool/src/lib.rs:270]: I think we're at risk of the deadlock described [in this HackMD](https://hackmd.io/ijQlqOdqSGaWmIo6zMVV-A?view#The-deadlock-in-Lighthouse) (see also https://github.com/sigp/lighthouse/pull/1562).


The `validity_filter` function can try to access the fork choice read lock:

https://github.com/sigp/lighthouse/blob/051c3e842fb5553d3ef3370560b74072c6e07cc6/beacon_node/beacon_chain/src/beacon_chain.rs#L2297

That means this function would behave the same as the `rest_api::validator::publish_attestations` function I describe in the HackMD.

In terms of a workaround, I'm not fully across this PR but I feel like the parallel functionality is most important for the second `map` fn (the one which calls `bron_kerbosch`). So perhaps the `filter` and first `map` could happen in a good ol' fashioned serial iter?

**Review by paulhauner — CHANGES_REQUESTED** [review]
Heyo, looking very impressive!

I had a look at just the Rayon code and I think there's a potential for a deadlock in there (see my comment).

File: beacon_node/operation_pool/src/lib.rs (hunk 7 above)
Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs:270]: Hmm, good point. I'm trying to understand how we can deadlock if we are only reading and never holding a write lock while spawning a Rayon task. I _think_ we are safe because we only have reads (item 2 from the HackMD) and not writes (item 1).

Reading the Rayon docs a bit more, I think how the deadlock happens is this:

1. Thread 1 obtains the write lock and then spawns a Rayon job on the Rayon thread pool and blocks waiting for it to complete.
2. Concurrently, another thread spawns a bunch of jobs on the Rayon thread pool which try to grab the read lock.
3. Rayon's pool "schedules" (="steals work such that") all of the reading jobs begin executing _but not_ the job spawned by thread 1 which is _required to release the lock_. Unlike `async` tasks in Tokio-land, there is no implicit yielding from Rayon jobs once they start executing.
4. Therefore we get a deadlock: all the threads in the pool are blocked waiting for a job that can only execute in the pool, which will never execute.

Without holding an exclusive lock/mutex while spawning I think we are safe. If all the jobs in the pool are just grabbing read locks and there is no other thread holding a write lock while it waits for a Rayon job to complete, then we are OK. Even if we have other (non-Rayon) threads grabbing the write lock, this is OK too, as they will be run concurrently by the OS, and the write lock will eventually be released, allowing the Rayon threads that are reading to make progress.

I went looking for places where we are already locking an `RwLock`/`Mutex` inside a Rayon job and actually the current version of the op pool uses `rayon::join` on two lazy iterators that evaluate the same `validity_filter` (and therefore obtain a read lock on fork choice from within the Rayon pool):

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/operation_pool/src/lib.rs#L307-L325

i.e. we are already doing this and it is not deadlocking

Additionally in Milhouse, we are recursively obtaining a `write` lock inside Rayon jobs here:

https://github.com/sigp/milhouse/blob/6c82029bcbc656a3cd423d403d7551974818d45d/src/tree.rs#L430-L433

Mac and I played around with that code trying to use upgradable locks and quickly hit deadlocks, which I think implies the current pattern is safe (see our attempts in https://github.com/sigp/milhouse/pull/25, https://github.com/sigp/milhouse/pull/29).

TL;DR: I think that Rayon is only dangerous is you _hold_ a lock while spawning jobs into the pool.

This was what I previously thought, but I'd never understood _why_ this was the case and thought it had something to do with thread-local storage, context switches and mutexes. I find the explanation above quite satisfactory and hope you do too!

Also, if we were to remove the `par_iter` we need a collect, as @GeemoCandama showed in 376c40845753f782bca5ab1cc910a6b734050c41. If you agree with my analysis I reckon we revert that commit :pray:

Thanks for reviewing!

Comment (paulhauner) [reply beacon_node/operation_pool/src/lib.rs:270]: > Without holding an exclusive lock/mutex while spawning I think we are safe.

I think we hold an exclusive lock whilst spawning here:

https://github.com/sigp/lighthouse/blob/2bc9115a94bfcf1269f83f0a9e7a1321b2a7f70c/beacon_node/beacon_chain/src/beacon_chain.rs#L1500-L1502

We take a write-lock on fork choice, then we call `on_block` which might read a state from the DB which might involve rayon-parallelised tree hashing.

So the deadlock would be:

1. We take the fork choice write lock in `on_block` but don't yet trigger tree hashing.
1. We run the `par_iter` in this PR, which fills all the rayon threads with functions waiting on the fork choice read lock.
1. `on_block` tries to start tree hashing but it can't start because all the rayon threads are busy.
1. We are deadlocked because we need `on_block` to finish so we can drop the fork choice write-lock.

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs:270]: Oh bugger, I think that could deadlock even with our current code then.

I think we need to choose between:

1. Never hold a lock while spawning Rayon tasks, or
2. Never obtain a lock in a Rayon task that could be held over a spawn

I have a feeling that (1) is better in some way, but need to sleep on it.

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs:270]: Actually does `on_block` load states? I can't see it at a quick glance. I also don't think loading a state does any tree hashing (although it will in tree-states land, so definitely something to keep an eye on).

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs:270]: Ah yeah, here:

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/beacon_chain/src/beacon_fork_choice_store.rs#L328-L334

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs:270]: If I get some time today or tomorrow I'll try writing a test with `hiatus` that deadlocks on `unstable`.

Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs:270]: After _a lot_ of mucking around I got block production & tree hashing to deadlock on `unstable`, sort of. The branch is here: https://github.com/michaelsproul/lighthouse/commit/5e253dee7bda9b22ae537af666122707aebd8cf7

It was really hard to get all the bits of line up, so there are 7 separate `hiatus` breakpoints. I also had to cheat and just force tree hashing in `import_block` because I ran out of patience to get the tree hashing and state load in `on_block` to trigger. The main problem is that the attestations were triggering the justified checkpoint update before I could trigger it with `on_block`. I think using the mainnet spec would make this a bit easier as we wouldn't have to justify on the epoch boundary (as we do with minimal).

I think this deadlock (on `unstable`) hasn't been observed in practice for several reasons:

1. It requires block production and block import to be running concurrently. This doesn't happen particularly often, but isn't _that_ unlikely.
2. It requires the balances cache in fork choice to miss. This almost never happens in recent versions of Lighthouse.
3. It requires a machine with ~2 cores, because we only spawn two op pool tasks using `rayon::join`. If there are more than 2 Rayon threads in the pool then the tree hashing can make progress and drop the fork choice lock, which unblocks the op pool tasks.

In my example you can see the deadlock happen with:

```
RAYON_NUM_THREADS=2 FORK_NAME=capella cargo test --release --features fork_from_env -- deadlock_block_production --nocapture
```

If you bump the threadpool size to 3 then the deadlock doesn't happen (due to condition 3).

```
RAYON_NUM_THREADS=3 FORK_NAME=capella cargo test --release --features fork_from_env -- deadlock_block_production --nocapture
```

I haven't tried deadlocking Geemo's branch yet, but intuitively we make the deadlock more likely by spawning more threads for the attestation data (condition 3 no longer applies). Still, we are protected by the unlikeliness of condition 1 & condition 2 occurring simultaneously.

I need to think about it more.

**Comment by michaelsproul** [comment]
Hey @GeemoCandama I think to play it safe lets remove all the Rayon calls and assess the performance impact. After going deep down the Rayon rabbit hole I don't think there's a completely safe way for us to use Rayon in the op pool at the moment (see https://github.com/sigp/lighthouse/issues/4952).

I think we should revisit parallelism in the op pool at a later date, maybe using a different dedicated thread pool, or Tokio. Hopefully #4925 buys us a little bit of wiggle room by reducing block production times overall.

**Comment by GeemoCandama** [comment]
Screenshot from 2024-01-23 00-32-18

**Comment by michaelsproul** [comment]
I did some benchmarking with `--subscribe-all-subnets` and attestation packing was taking 100-250ms regularly. I think we need the thread pool. I will add a non-Rayon thread pool when I have some time. I'll bump this from the v5.0.0 release, which needs to happen next week.

**Comment by michaelsproul** [comment]
Closing as stale.

We may still want to revisit this at some point.

I recently found an implementation of Bron-Kerbosch in `petgraph` which could be useful:

https://docs.rs/petgraph/0.8.2/petgraph/algo/maximal_cliques/fn.maximal_cliques.html
